### PR TITLE
Fixes for various edge cases

### PR DESCRIPTION
**Pull request by darrylmasson**
Mainly the ones that disabled the dispatcher after linked-mode tests with the NV.

### CODE REVIEW

**Review by petergaemers — APPROVED** [review]

File: dispatcher/MongoConnect.py [hunk 1]
```diff
@@ -628,9 +634,14 @@ def detector_ackd_command(self, detector, command):
...
-            # loop over doc['host'] because the 'acknowledged' field sometimes
-            # contains extra entries (such as the GPS trigger)
+        # we can't naively use everything in the hosts field, because we might be transitioning
+        # out of linked mode, and there might be "garbage" in the host list because someone
+        # didn't follow very clear instructions, and if a stop is issued to a host that doesn't
+        # exist, the dispatcher basically stops working
+        hosts_this_detector = set(list(self.latest_status[detector]['readers'].keys()) + list(self.latest_status[detector]['controller'].keys()))
+        hosts_in_doc = set(doc['host'])
+        # so we only loop over the intersection of this detector's hosts and the doc's hosts
+        for h in hosts_this_detector & hosts_in_doc:
```
Comment (jorana) [inline dispatcher/MongoConnect.py:644]: Just to be sure, this is one of the key fixes right? So we are going to filter out hosts that are not in the intercept of the sets for the latest status for this logical detector and the command queue for this logical detector.

You think it's also worth logging that we ignore hosts in this hosts-set:
```python
hosts_ignored = hosts_this_detector  ^ hosts_in_doc
if len(hosts_ignored):
   run_id = self.latest_status[detector]['number']
   self.log.warning(f'ignoring {hosts_ignored} for run {run_id}')
```

File: dispatcher/MongoConnect.py [hunk 2]
```diff
@@ -541,10 +541,16 @@ def send_command(self, command, hosts, user, detector, mode="", delay=0, force=F
...
         number = None
+        ls = self.latest_status[detector]
         if command == 'stop' and not self.detector_ackd_command(detector, 'stop'):
             self.log.error(f"{detector} hasn't ack'd its last stop, let's not flog a dead horse")
             if not force:
                 return 1
+        for host_list in hosts:
+            for h in host_list:
+                if h not in ls['readers'] or h not in ls['controller']:
+                    self.log.debug(f'Trying to issue a {command} to {detector}/{h}?')
```
[the remark this answers is not in the document]
Comment (darrylmasson) [reply dispatcher/MongoConnect.py:552]: It actually doesn't, because only `critical` from the python logger goes to the database.

File: dispatcher/MongoConnect.py [hunk 3]
```diff
@@ -541,10 +541,16 @@ def send_command(self, command, hosts, user, detector, mode="", delay=0, force=F
         Send this command to these hosts. If delay is set then wait that amount of time
         """
         number = None
+        ls = self.latest_status[detector]
         if command == 'stop' and not self.detector_ackd_command(detector, 'stop'):
             self.log.error(f"{detector} hasn't ack'd its last stop, let's not flog a dead horse")
             if not force:
                 return 1
+        for host_list in hosts:
```
[the remark this answers is not in the document]
Comment (darrylmasson) [reply dispatcher/MongoConnect.py:549]: Yes, it probably should

File: dispatcher/MongoConnect.py (hunk 1 above)
Comment (darrylmasson) [reply dispatcher/MongoConnect.py:644]: More logging never hurts

File: dispatcher/MongoConnect.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (darrylmasson) [reply dispatcher/MongoConnect.py:549]: Actually, it doesn't make that much difference. `detector_ackd_command` doesn't use any info from the desired host list here, but it would give us an extra log statement that could help.

File: dispatcher/MongoConnect.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jorana) [reply dispatcher/MongoConnect.py]: good catch

**Review by jorana — APPROVED** [review]
